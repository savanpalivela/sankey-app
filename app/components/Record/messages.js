/*
 * Record Messages
 *
 * This contains all the text for the Record component.
 */

import { defineMessages } from 'react-intl';

export const scope = 'app.components.Record';

export default defineMessages({
  header: {
    id: `${scope}.header`,
    defaultMessage: 'This is the Record component!',
  },
  income: {
    id: `${scope}.income`,
    defaultMessage: 'Income Records',
  },
  expense: {
    id: `${scope}.expense`,
    defaultMessage: 'Expense Records',
  },
});
