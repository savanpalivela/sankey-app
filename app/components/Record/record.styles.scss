@import '../../utils/styles/variables.scss';

.selectableRecord {
  margin: 10px 0;

  &:focus {
    outline: none;
  }

  &.active {
    background-color: $selected-att;
  }
}