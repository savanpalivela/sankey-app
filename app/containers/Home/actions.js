/*
 *
 * Home actions
 *
 */

import {
  DEFAULT_ACTION,
  GET_SANKEY_DATA,
  GET_SANKEY_DATA_SUCCESS,
  GET_SANKEY_DATA_ERROR,
  UPDATE_RECORD_DATA,
  REMOVE_ATT_RECORD_DATA,
} from './constants';

const defaultAction = () => ({
  type: DEFAULT_ACTION,
});

const editRecordValuesAction = payload => ({
  type: UPDATE_RECORD_DATA,
  payload,
});

const loadSankeyData = () => ({
  type: GET_SANKEY_DATA,
});

const setSankeyData = payload => ({
  type: GET_SANKEY_DATA_SUCCESS,
  payload,
});

const setSankeyDataError = payload => ({
  type: GET_SANKEY_DATA_ERROR,
  payload,
});

const removeRecordAttributeAction = payload => ({
  type: REMOVE_ATT_RECORD_DATA,
  payload,
});

export {
  defaultAction,
  loadSankeyData,
  setSankeyData,
  setSankeyDataError,
  editRecordValuesAction,
  removeRecordAttributeAction,
};
