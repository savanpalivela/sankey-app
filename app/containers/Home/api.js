const URL = 'https://5ee2faa18b27f300160951b4.mockapi.io/api/data/';

const fetchSankeyData = async () => {
  const response = await Promise.all([
    fetch(`${URL}income`),
    fetch(`${URL}expense`),
  ]);
  // if (response[0].status >= 400 || response[1].status >= 400) {
  //   throw new Error(data.errors);
  // }
  const data = await Promise.all([response[0].json(), response[1].json()]);
  return {
    income: data[0],
    expense: data[1],
  };
};

export { fetchSankeyData };
