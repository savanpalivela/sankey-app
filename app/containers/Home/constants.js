/*
 *
 * Home constants
 *
 */

export const DEFAULT_ACTION = 'app/Home/DEFAULT_ACTION';
export const GET_SANKEY_DATA = 'GET_SANKEY_DATA';
export const GET_SANKEY_DATA_SUCCESS = 'GET_SANKEY_DATA_SUCCESS';
export const GET_SANKEY_DATA_ERROR = 'GET_SANKEY_DATA_ERROR';
