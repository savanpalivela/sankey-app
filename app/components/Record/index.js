/**
 *
 * Record
 *
 */

import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
// import styled from 'styled-components';

import { FormattedMessage } from 'react-intl';
import { Plus } from 'react-bootstrap-icons';
import messages from './messages';
import RecordAttribute from '../RecordAttribute';

function Record({
  title,
  data,
  handleValueChange,
  handleRemoveAttRecord,
  handleAddAttRecord,
  attributeKeyList,
}) {
  const [showAddEl, setShowAddEl] = useState(false);
  const [unusedAttributes, setUnusedAttributes] = useState(attributeKeyList);
  const [newRecordAttObj, setNewRecordAttObj] = useState({ key: '', value: 0 });

  useEffect(() => {
    if (data) {
      const usedAttKeys = data.map(i => i.key);
      const newUnusedAttributes = attributeKeyList.filter(
        x => !usedAttKeys.includes(x),
      );
      setUnusedAttributes(() => newUnusedAttributes);
    }
  }, [data]);

  const handleAddAttClick = () => {
    handleAddAttRecord(newRecordAttObj);
    setNewRecordAttObj(() => ({ key: '', value: 0 }));
    setShowAddEl(false);
  };

  const addRecordAttEl = () => {
    if (showAddEl) {
      return (
        <div className="row">
          <div className="col-sm-5">
            <select
              className="form-control"
              onChange={e => {
                setNewRecordAttObj({
                  ...newRecordAttObj,
                  key:
                    e.currentTarget.options[e.currentTarget.selectedIndex]
                      .value,
                });
              }}
            >
              <FormattedMessage {...messages.defaultSelect}>
                {message => <option value="">{message}</option>}
              </FormattedMessage>
              {unusedAttributes.map((item, i) => {
                const key = `keys-${title}-${i}`;
                return (
                  <FormattedMessage {...messages[item]} key={key}>
                    {message => <option value={item}>{message}</option>}
                  </FormattedMessage>
                );
              })}
            </select>
          </div>
          <div className="col-sm-5">
            <input
              className="form-control"
              type="number"
              value={newRecordAttObj.value}
              onChange={e => {
                setNewRecordAttObj({
                  ...newRecordAttObj,
                  value: parseInt(e.currentTarget.value, 10),
                });
              }}
            />
          </div>
          <div className="col-sm-2">
            <button
              disabled={
                newRecordAttObj.key === '' || newRecordAttObj.value === 0
              }
              className="btn btn-light"
              type="button"
              onClick={handleAddAttClick}
            >
              <FormattedMessage {...messages.addRecordAtt} />
            </button>
          </div>
        </div>
      );
    }
    return null;
  };

  return (
    <div className="card">
      <div className="card-header">
        <FormattedMessage {...messages[title]} />
      </div>
      <div className="card-body">
        {data &&
          data.length > -1 &&
          data.map((item, i) => (
            <RecordAttribute
              key={item.key}
              keyText={item.key}
              value={item.value}
              handleValueChange={val => handleValueChange(item.key, val)}
              handleRemoveAttRecord={() => handleRemoveAttRecord(i)}
            />
          ))}
        {addRecordAttEl()}
        {unusedAttributes && unusedAttributes.length === 0 ? null : (
          <button
            className="btn btn-light"
            type="button"
            onClick={() => {
              setShowAddEl(true);
            }}
          >
            <Plus size={24} />
          </button>
        )}
      </div>
    </div>
  );
}

Record.propTypes = {
  title: PropTypes.string.isRequired,
  data: PropTypes.arrayOf(PropTypes.object).isRequired,
  handleValueChange: PropTypes.func.isRequired,
  handleRemoveAttRecord: PropTypes.func.isRequired,
  handleAddAttRecord: PropTypes.func.isRequired,
  attributeKeyList: PropTypes.arrayOf(PropTypes.string).isRequired,
};

export default Record;
