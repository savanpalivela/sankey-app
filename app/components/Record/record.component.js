/**
 *
 * Record
 *
 */

import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';

import { FormattedMessage } from 'react-intl';
import { Plus } from 'react-bootstrap-icons';
import messages from './record.messages';
import RecordAttribute from '../RecordAttribute';
import styles from './record.styles.scss';
function Record({
  title,
  data,
  handleValueChange,
  handleRemoveAttRecord,
  handleAddAttRecord,
  attributeKeyList,
  onRowSelect,
  selectedAttrubiteIndex,
  enableAttSelection,
  titleSecondary,
}) {
  const [showAddEl, setShowAddEl] = useState(false);
  const [unusedAttributes, setUnusedAttributes] = useState(attributeKeyList);
  const [newRecordAttObj, setNewRecordAttObj] = useState({ key: '', value: 0 });

  useEffect(() => {
    if (data) {
      const usedAttKeys = data.map(i => i.key);
      const newUnusedAttributes = attributeKeyList.filter(
        x => !usedAttKeys.includes(x),
      );
      setUnusedAttributes(() => newUnusedAttributes);
    }
  }, [data]);

  const handleAddAttClick = () => {
    handleAddAttRecord(newRecordAttObj);
    setNewRecordAttObj(() => ({ key: '', value: 0 }));
    setShowAddEl(false);
  };

  const getRecordEl = (item, i) => {
    const recordEl = (
      <RecordAttribute
        keyText={item.key}
        value={item.value}
        handleValueChange={val => handleValueChange(i, val)}
        handleRemoveAttRecord={() => handleRemoveAttRecord(i)}
      />
    );

    if (enableAttSelection) {
      return (
        <div
          role="button"
          onClick={() => {
            onRowSelect(i);
          }}
          onKeyDown={() => {
            onRowSelect(i);
          }}
          tabIndex={0}
          className={`${
            selectedAttrubiteIndex === i ? styles.active : ''
          } card ${styles.selectableRecord}`}
        >
          {recordEl}
        </div>
      );
    }
    return recordEl;
  };

  // Add record Attribute Element
  const addRecordAttEl = () =>
    showAddEl && (
      <div className="row" data-testid="add-attribute-el">
        <div className="col-sm-5">
          <select
            className="form-control"
            onChange={e => {
              setNewRecordAttObj({
                ...newRecordAttObj,
                key:
                  e.currentTarget.options[e.currentTarget.selectedIndex].value,
              });
            }}
          >
            <FormattedMessage {...messages.defaultSelect}>
              {message => <option value="">{message}</option>}
            </FormattedMessage>
            {unusedAttributes.map(item => {
              const key = `keys-${title}-${item}`;
              return (
                <FormattedMessage {...messages[item]} key={key}>
                  {message => <option value={item}>{message}</option>}
                </FormattedMessage>
              );
            })}
          </select>
        </div>
        <div className="col-sm-5">
          <input
            className="form-control"
            type="number"
            value={newRecordAttObj.value}
            onChange={e => {
              setNewRecordAttObj({
                ...newRecordAttObj,
                value: parseInt(e.currentTarget.value, 10),
              });
            }}
          />
        </div>
        <div className="col-sm-2">
          <button
            disabled={newRecordAttObj.key === '' || newRecordAttObj.value === 0}
            className="btn btn-light"
            type="button"
            data-testid="add-attribute-btn"
            onClick={handleAddAttClick}
          >
            <FormattedMessage {...messages.addRecordAtt} />
          </button>
        </div>
      </div>
    );

  return (
    <div className="card">
      <div className="card-header">
        <FormattedMessage {...messages[title]} />
        {titleSecondary && (
          <span>
            &nbsp;Mapped to&nbsp;
            <strong>
              <FormattedMessage {...messages[titleSecondary]} />
            </strong>
          </span>
        )}
      </div>
      <div className="card-body" data-testid="record-main">
        {data &&
          data.length > -1 &&
          data.map((item, i) => (
            <div key={item.key}>{getRecordEl(item, i)}</div>
          ))}
        {addRecordAttEl()}
        {unusedAttributes && unusedAttributes.length === 0 ? null : (
          <button
            data-testid="enable-add-btn"
            className="btn btn-light"
            type="button"
            onClick={() => {
              setShowAddEl(true);
            }}
          >
            <Plus size={24} />
          </button>
        )}
      </div>
    </div>
  );
}

Record.propTypes = {
  title: PropTypes.string.isRequired,
  titleSecondary: PropTypes.string,
  data: PropTypes.arrayOf(PropTypes.object).isRequired,
  handleValueChange: PropTypes.func.isRequired,
  handleRemoveAttRecord: PropTypes.func.isRequired,
  handleAddAttRecord: PropTypes.func.isRequired,
  attributeKeyList: PropTypes.arrayOf(PropTypes.string).isRequired,
  onRowSelect: PropTypes.func,
  selectedAttrubiteIndex: PropTypes.number,
  enableAttSelection: PropTypes.bool,
};

Record.defaultProps = {
  onRowSelect: () => {},
  selectedAttrubiteIndex: -1,
  enableAttSelection: false,
  titleSecondary: '',
};

export default Record;
