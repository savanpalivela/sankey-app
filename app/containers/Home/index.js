/**
 *
 * Home
 *
 */

import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Helmet } from 'react-helmet';
import { FormattedMessage } from 'react-intl';
import { createStructuredSelector } from 'reselect';
import { compose } from 'redux';

import { useInjectSaga } from 'utils/injectSaga';
import { useInjectReducer } from 'utils/injectReducer';
import {
  makeSelectNode,
  makeSelectLink,
  makeSelectIsLoaded,
  makeSelectIncome,
  makeSelectExpense,
} from './selectors';
import reducer from './reducer';
import saga from './saga';
import messages from './messages';
import SankeyChart from '../../components/Sankey';
import Record from '../../components/Record';
import {
  loadSankeyData,
  editRecordValuesAction,
  removeRecordAttributeAction,
  addRecordAttributeAction,
} from './actions';
import { INCOME_RECORD_ATT_KEYS, EXPENSE_RECORD_ATT_KEYS } from './constants';
import styles from './styles.scss';
export function Home({
  loadData,
  nodes,
  links,
  isLoaded,
  income,
  expense,
  editRecordValue,
  removeRecordAttribute,
  addRecordAttribute,
}) {
  useInjectReducer({ key: 'home', reducer });
  useInjectSaga({ key: 'home', saga });
  useEffect(() => {
    if (!isLoaded) {
      loadData();
    }
  });

  const handleValueChange = (recordType, key, value) => {
    editRecordValue(recordType, key, value);
  };

  return (
    <div>
      <Helmet>
        <title>Home</title>
        <meta name="description" content="Description of Home" />
      </Helmet>
      <h3 className={styles.pageTitle}>
        <FormattedMessage {...messages.header} />
      </h3>
      <SankeyChart nodes={nodes} links={links} />
      <div className="row">
        <div className="col-sm-6">
          <Record
            title="income"
            data={income}
            attributeKeyList={INCOME_RECORD_ATT_KEYS}
            handleValueChange={(key, val) => {
              handleValueChange('income', key, val);
            }}
            handleRemoveAttRecord={index => {
              removeRecordAttribute('income', index);
            }}
            handleAddAttRecord={attObj => {
              addRecordAttribute('income', attObj);
            }}
          />
        </div>
        <div className="col-sm-6">
          <Record
            title="expense"
            data={expense}
            attributeKeyList={EXPENSE_RECORD_ATT_KEYS}
            handleValueChange={(key, val) => {
              handleValueChange('expense', key, val);
            }}
            handleRemoveAttRecord={index => {
              removeRecordAttribute('expense', index);
            }}
            handleAddAttRecord={attObj => {
              addRecordAttribute('expense', attObj);
            }}
          />
        </div>
      </div>
    </div>
  );
}

Home.propTypes = {
  nodes: PropTypes.arrayOf(PropTypes.object).isRequired,
  links: PropTypes.arrayOf(PropTypes.object).isRequired,
  expense: PropTypes.arrayOf(PropTypes.object).isRequired,
  income: PropTypes.arrayOf(PropTypes.object).isRequired,
  isLoaded: PropTypes.bool.isRequired,
  loadData: PropTypes.func.isRequired,
  removeRecordAttribute: PropTypes.func.isRequired,
  addRecordAttribute: PropTypes.func.isRequired,
  editRecordValue: PropTypes.func.isRequired,
};

const mapStateToProps = createStructuredSelector({
  nodes: makeSelectNode(),
  links: makeSelectLink(),
  income: makeSelectIncome(),
  expense: makeSelectExpense(),
  isLoaded: makeSelectIsLoaded(),
});

function mapDispatchToProps(dispatch) {
  return {
    loadData: () => {
      dispatch(loadSankeyData());
    },
    editRecordValue: (recordType, key, value) => {
      const payload = {
        recordType,
        key,
        value,
      };
      dispatch(editRecordValuesAction(payload));
    },
    removeRecordAttribute: (recordType, index) => {
      const payload = {
        recordType,
        index,
      };
      dispatch(removeRecordAttributeAction(payload));
    },
    addRecordAttribute: (recordType, attObj) => {
      const payload = {
        recordType,
        attObj,
      };
      dispatch(addRecordAttributeAction(payload));
    },
  };
}

const withConnect = connect(
  mapStateToProps,
  mapDispatchToProps,
);

export default compose(withConnect)(Home);
