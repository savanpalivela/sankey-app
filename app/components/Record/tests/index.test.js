/**
 *
 * Tests for Record
 *
 * @see https://github.com/react-boilerplate/react-boilerplate/tree/master/docs/testing
 *
 */

import React from 'react';
import { render } from 'react-testing-library';
import { IntlProvider } from 'react-intl';
import { INCOME_5 } from '../../../../internals/mocks/data';
import { INCOME_RECORD_ATT_KEYS } from '../../../containers/Home/constants';
// import 'jest-dom/extend-expect'; // add some helpful assertions

import Record from '../index';
import { DEFAULT_LOCALE } from '../../../i18n';

describe('<Record />', () => {
  it('Expect to not log errors in console', () => {
    const spy = jest.spyOn(global.console, 'error');
    render(
      <IntlProvider locale={DEFAULT_LOCALE}>
        <Record
          title="income"
          data={INCOME_5}
          attributeKeyList={INCOME_RECORD_ATT_KEYS}
          handleValueChange={() => {}}
          handleRemoveAttRecord={() => {}}
          handleAddAttRecord={() => {}}
        />
      </IntlProvider>,
    );
    expect(spy).not.toHaveBeenCalled();
  });
});
