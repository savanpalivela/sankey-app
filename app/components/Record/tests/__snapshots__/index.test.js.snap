// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Record /> should render and match the snapshot 1`] = `
<Record
  attributeKeyList={
    Array [
      "services",
      "recurring",
      "project",
      "consultation",
      "interest",
    ]
  }
  data={
    Array [
      Object {
        "key": "services",
        "value": 500,
      },
      Object {
        "key": "recurring",
        "value": 540,
      },
      Object {
        "key": "project",
        "value": 1000,
      },
      Object {
        "key": "interest",
        "value": 300,
      },
    ]
  }
  enableAttSelection={false}
  handleAddAttRecord={[Function]}
  handleRemoveAttRecord={[Function]}
  handleValueChange={[Function]}
  onRowSelect={[Function]}
  selectedAttrubiteIndex={-1}
  title="income"
  titleSecondary=""
/>
`;
